tty/usbacm/usbacm.c

```diff
@@ -302,62 +367,114 @@ static usbacm_dev_t *usbacm_devAlloc(void)
 }
 
 
-static int _usbacm_open(usbacm_dev_t *dev, int flags)
+static int _usbacm_urbsAlloc(usbacm_dev_t *dev)
 {
 	int i;
-	fifo_t *fifo;
 
-	if (dev->flags != 0)
-		return -EIO;
+	dev->urbctrl = usb_urbAlloc(dev->pipeCtrl, NULL, usb_dir_out, 0, usb_transfer_control);
+	if (dev->urbctrl < 0) {
+		return -1;
+	}
 
-	if (dev->fifo == NULL) {
-		if ((fifo = malloc(sizeof(fifo_t) + RX_FIFO_SIZE)) == NULL)
-			return -ENOMEM;
+	for (i = 0; i < USBACM_N_URBS; i++) {
+		dev->urbs[i] = usb_urbAlloc(dev->pipeBulkIN, NULL, usb_dir_in, USBACM_BULK_SZ, usb_transfer_bulk);
+		if (dev->urbs[i] < 0) {
+			while (--i >= 0) {
+				usb_urbFree(dev->pipeBulkIN, dev->urbs[i]);
+				dev->urbs[i] = 0;
+			}
+			usb_urbFree(dev->pipeCtrl, dev->urbctrl);
+			return -1;
+		}
+	}
 
-		fifo_init(fifo, RX_FIFO_SIZE);
+	return 0;
+}
 
-		if (mutexCreate(&dev->fifoLock) != 0) {
-			free(fifo);
-			return -ENOMEM;
-		}
 
-		if (condCreate(&dev->fifoEmpty) != 0) {
-			resourceDestroy(dev->fifoLock);
-			free(fifo);
-			return -ENOMEM;
-		}
-		dev->fifo = fifo;
-	}
+static int _usbacm_openNonblock(usbacm_dev_t *dev)
+{
+	fifo_t *fifo;
+	int ret = 0;
 
-	dev->flags = flags;
-	for (i = 0; i < USBACM_N_URBS; i++)
-		dev->urbs[i] = usb_urbAlloc(dev->pipeBulkIN, NULL, usb_dir_in, USBACM_BULK_SZ, usb_transfer_bulk);
+	if ((fifo = malloc(sizeof(fifo_t) + RX_FIFO_SIZE)) == NULL)
+		return -ENOMEM;
 
-	for (i = 0; i < USBACM_N_URBS; i++)
-		usb_transferAsync(dev->pipeBulkIN, dev->urbs[i], USBACM_BULK_SZ);
+	fifo_init(fifo, RX_FIFO_SIZE);
 
-	mutexUnlock(usbacm_common.lock);
+	mutexLock(dev->rxLock);
+	dev->fifo = fifo;
+	dev->rxrunning = 1;
 
-	if (usbacm_init(dev) != EOK) {
-		fprintf(stderr, "usbacm: Init failed\n");
-		resourceDestroy(dev->fifoLock);
-		resourceDestroy(dev->fifoEmpty);
-		free(dev->fifo);
-		return -EINVAL;
+	if (usbacm_rxStart(dev) != 0) {
+		free(fifo);
+		dev->fifo = 0;
+		dev->rxrunning = 0;
```

Comment:
Hmm, condSignal not needed here?

Reply:
No, this is a nonblocking case - no waiting on the conditional.